### ISSUE TITLE
Allow custom waveFormat for WasapiCapture without calling SetupWaveFormat()

### ISSUE THREAD

**Issue by DanielGilbert** [filoe/cscore#277]
Hi,

_Dell/Alienware/Realtek are driving me nuts. This is the biggest **** I've ever encountered, and explains a lot of "I cannot initialize my device, app crashes" errors._

I'm using WasapiCapture to record music from the default audio device. When I'm using my internal, out-of-the-box soundcard in my Alienware Laptop, I get this error:

> IAudioClient::Initialize caused an error: 0x88890008, "Unknown HRESULT".

I was able to pin-point it down to the default MixFormat the device is reporting. It looks like this:

`{ChannelsAvailable: 8|SampleRate: 44100|Bps: 1411200|BlockAlign: 32|BitsPerSample: 32|Encoding: Extensible|SubFormat: 00000003-0000-0010-8000-00aa00389b71|ChannelMask: SpeakerFrontLeft, SpeakerFrontRight, SpeakerFrontCenter, SpeakerLowFrequency, SpeakerBackLeft, SpeakerBackRight, SpeakerSideLeft, SpeakerSideRight}`

You might notice that it says it has 8 channels - I can confirm that this is not true, and  has to do with some built-in virtual enhancement stuff. So I changed the Channel and Mask properties in **the debugger right before** the Initialization call, and everything works.

However, when I supply my WaveFormat structure to the WasapiCapture constructor, it crashes again. This is related to [InitializeInternal()](https://github.com/filoe/cscore/blob/master/CSCore/SoundIn/WasapiCapture.cs#L351), where the driver gets asked if it supports the given WaveFormat - which it denies when I provide my "custom" format, and instead supposes the not working 8 channel config, which CSCore happily accepts.

**tl;dr:** Would it be possible to add a "trust me, I'm an engineer" flag for the initialization routine, which prevents the WaveFormat double check?

**Comment by DanielGilbert** [comment]
Ah, silly me: I forgot to use WaveFormatExtensible instead of WaveFormat as a parameter to the constructor when dealing with > 2 channels. If I provide a WaveFormatExtensible, SetupWaveFormat gets the correct response from the driver, and stuff works.  Still, I need to change the channels to "2" instead of the default reported mix format.

Case closed.
